### PR TITLE
Add python `nested_tensor` and `as_nested_tensor` constructors in `torch.nested`

### PR DESCRIPTION
**Pull request by mikaylagawarecki**
Remove `torch.nested_tensor` which has erroneous behavior wrt gradients (could be either leaf or not leaf). Introduce `torch.nested.nested_tensor` and `torch.nested.as_nested_tensor` in the vein of `torch.tensor` and `torch.as_tensor`. Done in nested `__init__.py` for now but can move to pybind in future (when we want to load from numpy/nested lists ). 

Discussed offline with @cpuhrsch and pybind constructor (https://github.com/pytorch/pytorch/pull/85536) was more gnarly than expected, so we can move to that when we do need loading from numpy etc.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #85593

Differential Revision: [D39806622](https://our.internmc.facebook.com/intern/diff/D39806622)

### CODE REVIEW

**Comment by mikaylagawarecki** [comment]
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39806622).

**Comment by mikaylagawarecki** [comment]
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39806622).

File: torch/nested/__init__.py [hunk 1]
```diff
@@ -1,11 +1,89 @@
...
+        >>> b = torch.arange(5, dtype=torch.float, requires_grad=True)
+        >>> nt = torch.nested.nested_tensor([a, b], requires_grad=True)
+        >>> nt.is_leaf
+        True
+    """
+    if not isinstance(tensor_list, list) or any([not torch.is_tensor(t) for t in tensor_list]):
+        raise TypeError("nested_tensor(): Expected first argument to be a list of tensors ")
+    new_data = []
+    for t in tensor_list:
+        new_data.append(t.clone().detach())
```
Comment (drisspg) [inline torch/nested/__init__.py:48]: Small performance hit by having to clone but I think it is worth it in order to do this all in python

**Review by drisspg — APPROVED** [review]
Didn't look at all the call sites of `torch.nested.nested..` but looks good. @cpuhrsch do you have issue with the two functions created at the python level calling into the same native function

File: torch/nested/__init__.py (hunk 1 above)
Comment (cpuhrsch) [reply torch/nested/__init__.py:48]: Why do we have to clone?

**Review by cpuhrsch — APPROVED** [review]

File: torch/nested/__init__.py (hunk 1 above)
Comment (mikaylagawarecki) [reply torch/nested/__init__.py:48]: Should it just be `t.detach()`?

Comment (cpuhrsch) [reply torch/nested/__init__.py:48]: Yeah I think that's enough

Comment (mikaylagawarecki) [reply torch/nested/__init__.py:48]: Ah I see will change, I understand that `cat` in NestedTensorUtils:wrap_tensor_node will ultimately create a copy, but in practice I've never seen `detach` used without `clone`

**Comment by mikaylagawarecki** [comment]
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39806622).

File: torch/nested/__init__.py (hunk 1 above)
Comment (drisspg) [reply torch/nested/__init__.py]: In the readme: 
```
# NOTE: Previously, if we change the tensor metadata (e.g. sizes / strides /
# storage / storage_offset) of a tensor created from `detach()`, those metadata
# in the original tensor will also be updated. However, the new behavior is that
# those metadata changes to the detached tensor will not update the original tensor
# anymore, and in the `detach()` function we need to set `allow_tensor_metadata_change_`
# to false to make such changes explicitly illegal, in order to prevent users from
# changing metadata of the detached tensor and expecting the original tensor to also
# be updated.
```
I think the concern is that we don't want the detach to mutate the original tensor but like you said since wrap_tensor makes a copy I think in this case it is okay.

Comment (mikaylagawarecki) [reply torch/nested/__init__.py]: Ah gotcha thanks Driss :)

**Comment by mikaylagawarecki** [comment]
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39806622).

**Comment by mikaylagawarecki** [comment]
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39806622).
